Statically typed CloudFormation Templates

rosenhouse opened the pull request:
A draft, and not feature-complete.  Feedback appreciated.

Context from my original issue #393:
>I'm trying to programmatically generate CloudFormation templates.
>
>`cloudformation.CreateStackInput` has its `TemplateBody` field typed as a `*string`.  I'd like to be able to build up a strongly-typed Template in memory, call `json.Marshal()` and use the result as the `TemplateBody`.

I've roughly followed [the approach outlined](https://github.com/aws/aws-sdk-go/issues/393#issuecomment-144875078) by @jasdel .  One difference is that the intrinsic `Fn::x`  and `Ref` functions are modeled as Go functions, not types.  It makes the usage syntax a bit cleaner:
```go
FnJoin("\n", "Stack ID is:", Ref("AWS::StackId") )
```
instead of 
```go
FnJoin{ "\n", []interface{}{ "Stack ID is:", Ref{ "AWS::StackId" } } }
```
But perhaps there are some drawbacks I haven't considered.  And regardless, you'll probably want to dot-import this package.

What's missing:
- Types for [all the different resources](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-template-resource-type-ref.html)
- Types for [all the different resource properties](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-product-property-reference.html)
- [Resource attributes](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-product-attribute-reference.html) including [`DependsOn`](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-dependson.html) which shows up a lot with VPCs

It would be great to auto-generate the resource and resource-property types somehow.

But I wanted to get feedback before I invested too much more time.

jasdel commented:
Hi @rosenhouse thanks a lot for taking the time and putting this preliminary PR. I'm looking for predefined models of the template shapes to see if there is something we can use to generate code from.  In the mean time I'll review the PR and get back to you with feedback.

Using helper functions instead of structs for `Fn::x` should work just as well too.

rosenhouse commented:
@jasdel: Any update here?   Would you be open to merging this PR as it currently stands?  We could follow up later on with the generated resource & resource-property structs.

jasdel commented on service/cloudformation/template/template_test.go:
```diff
@@ -0,0 +1,118 @@
+	assertEquivalentJSON(t, expectedJSON, string(actual))
+}
+
+func assertEquivalentJSON(t *testing.T, expected, actual string) {
```
Can assertEquivalentJSON be replaced with `awstesting.AssertJSON`? , they look similar.

jasdel commented on service/cloudformation/template/template.go:
```diff
@@ -0,0 +1,69 @@
+package template
```
Would you mind adding a package doc header here?

jasdel commented:
Thanks a lot for your hard work getting this PR together @rosenhouse. It looks good and I think we can pull it in.

I agree being able to generate the code would be the best plan, especially for the resources and property resources. The .NET SDK uses a JSON scheme for their Visual Studio's CloudFormation Template generator wizard. I'm investigating if it is possible for the Go SDK to reuse the scheme for our resource and resource-property structs.

rosenhouse commented:
Thanks for the review.  I've rebased against `master`, replaced my `assertEquivalentJSON` with `awstesting.AssertJSON` and added a package doc header.

jasdel commented:
Looks great, thanks, and I merged the change in. This will be a great place to get started building templates programmatically. I'll update #393 when more information is available about auto generating the resources and resource-properties.

rosenhouse commented:
Great, thank you!
